How to Develop and Deploy Fast Applications with Vite JS (#7136)
section-engineering-education/engineering-education: content/articles/develop-and-deploy-fast-apps-with-vite-js/index.md

Vite comes in when old tools like Webpack and Rollup have started to feel slower. The problem that Vite is trying to solve is the feedback loop speed as it is super fast.

Vite aims to provide modern web projects with speedy apps and a seamless developer experience. This is done in two ways:
- During development, code is not bundled like when using Webpack - Instead, it supports ECMAScript 6 module (ESM). ESM support allows browsers to fetch modules when importing them. Vite uses ESM to eliminate the build step while in development fully. This is because it is supported by all recent browsers such as Chrome, Edge and Firefox to ensure ESM support.
- Code and libraries that need to be imported have to be compiled - Vite achieves this with the help of `esbuild`, a fast Go-based tool. ESbuild pre-bundles dependencies are faster than JavaScript bundles (WebPack).
- Vite uses Rollup to bundle your project during production - The development process is much smoother since this compilation phase isn't necessary until the code is ready for production or testing.

### How Vite works

> Vite immediately launches the server. It uses esbuild to pre-bundle the dependencies that don't change frequently. Vite analyzes the code elements ought to be processed via route-based code splitting. Everything does not have to be bundled.

Reviewer: ```suggestion
Vite immediately launches the server. It uses esbuild to pre-bundle the dependencies that don't change frequently. Vite analyzes the code elements that ought to be processed via route-based code splitting. Everything does not have to be bundled.

Vite then delivers the code using native ESM support in modern browsers. This lets the browser take the job of bundling and development. But for production, Vite uses Rollup to bundle and implement other performance optimizations.

### Features of Vite
Vite has features that help you know why you should use it and why you need it:
- Vite is compatible with ECMAScript modules.

> - [Vite offers Hot Module Replacement (HMR)](https://vitejs.dev/guide/api-hmr.html#hot-accept-cb).

Reviewer: ```suggestion
-  Vite offers [Hot Module Replacement (HMR)](https://vitejs.dev/guide/api-hmr.html#hot-accept-cb).
```

> - [Vite offers Hot Module Replacement (HMR)](https://vitejs.dev/guide/api-hmr.html#hot-accept-cb).

Reviewer: ```suggestion
-  Vite offers [Hot Module Replacement (HMR)](https://vitejs.dev/guide/api-hmr.html#hot-accept-cb).
```

- Vite has faster dependency pre-building - It uses esbuild instead of Rollup for dependency pre-building, and es-build is faster.
- It has typescript support.
- It has CSS pre-processor support.
- Vite has automatic CSS code-splitting - It makes your site much faster as it only loads the CSS code needed for the specific part of the application.

### Develop and deploy fast applications with Vite
So let's learn by doing. Let us develop and deploy a Vanilla.js Vite application.

#### Vite Installation
To get started with Vite, we will need Node.js and npm installed on your computer. Open the terminal in your VS Code and run the command below:

`$ npm init vite@latest`

![Project name](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/project-name.jpg)

Afterwards, enter the project name and continue. We will then choose the JavaScript framework to use. We will use vanilla.js in this article. You'll see it has scaffold in your project. Then you can follow the instructions that you see.

The instructions include:

![Vanilla.js](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/vanilla-js.jpg)

> We will cd into our project and run the npm install. This means changing the directory of our project and installing all the dependencies.

Reviewer: What does it mean to "cd" into a project, kindly rephrase it in a better way.

> We will cd into our project and run the npm install. This means changing the directory of our project and installing all the dependencies.

Reviewer: What does it mean to "cd" into a project, kindly rephrase it in a better way.

![Instructions](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/instructions.jpg)

> Before running the `vite-project`, this is its structure after installing dependencies. We have a front-end structure with some `node modules` for our dependencies. We have a default `index.html` file. The CSS file gets imported at the `main.js` file.

Reviewer: What is its structure?. Are you referring to the image above?

> Before running the `vite-project`, this is its structure after installing dependencies. We have a front-end structure with some `node modules` for our dependencies. We have a default `index.html` file. The CSS file gets imported at the `main.js` file.

Author: I am referring to the next image.  This is a tutorial guiding readers on the Vite installation process.

![Installed dependencies](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/installed-dependencies.jpg)

The project structure is pretty simple. Let us look at what the `package.json` file contains. It has three scripts that we're going to use:

> - dev script - start our local folder.
> - build script - bundle our project for production.
> - preview script - look at our local build in our local environment.

Reviewer: ```suggestion
- dev script - We use it to start our local folder.
- build script - It is used to bundle our project for production.
- preview script - It is used to look at our local build in our local environment.
```

![Package.json](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/package-json.jpg)

> Since we have installed the dependencies, let's run `npm run dev`. Then, open the project at [localhost 3000](http://localhost:3000/). Click the link, and it will open on your browser.

Reviewer: State why we need to run the command `npm run dev`

> Since we have installed the dependencies, let's run `npm run dev`. Then, open the project at [localhost 3000](http://localhost:3000/). Click the link, and it will open on your browser.

Reviewer: State why we need to run the command `npm run dev`

Congratulations! You now have Vanilla.js and Vite running on your local server. You will notice the `hmr update` running when you edit files and changes appear instantly on the browser.

![Launch the project on the browser](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/npm-run.jpg)

![Vite project](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/vite-project.jpg)

To begin building or developing the project, run `npm run build`. The project is compiled into a `dist` folder containing the JavaScript and CSS files when that command is executed. Both files appear to be hashed.

![Dist](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/dist.jpg)

We execute the command `npm run preview` to preview the project. It will serve as a preview on [localhost 4173](http://localhost:4173/). So, 3000 is your build server, and 4173 is the preview for our build.

![Preview](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/preview.jpg)

> #### Build a JavaScrip app to learn more about Vite.

Reviewer: ```suggestion
#### Build a JavaScript app to learn more about Vite.
```

> #### Build a JavaScrip app to learn more about Vite.

Reviewer: ```suggestion
#### Build a JavaScript app to learn more about Vite.
```

Let us create a simple JavaScript application to learn more about Vite. We will create a file in our project's folder called `index.md`. The markdown file will have some data for us to load.

File `index.md`:

![Markdown file](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/markdown-cop.jpg)

> One of the awesome features of Vite is that you can easily import data into your JavaScript. We will need to install markdownIt `npm install markdown-it`. This will let us process text into markdown directly in my application.

Reviewer: ```suggestion
One of the awesome features of Vite is that you can easily import data into your JavaScript application. We will need to install markdownIt using the command  `npm install markdown-it`. This will let us process text into markdown directly in my application.

![MarkdownIt](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/markdown-it.jpg)

> Now we can import the library into our `main.js` file. You can not only import modules but data as well.

Reviewer: ```suggestion
Now we can import the library into our `main.js` file. We cannot only import modules but data as well.

> Now we can import the library into our `main.js` file. You can not only import modules but data as well.
> `import markdown-it from 'markdown-it'`

Reviewer: ```suggestion
Now we can import the library into our `main.js` file. We cannot only import modules but data as well. We use the command `import markdown-it from 'markdown-it'`  to do this
```

`import copy from './index.md?raw'`, the added a question mark with the value of raw helps to import it as text. Then this code will render our data in the app `markdownIt().render(copy)`. This is our `main.js` file:

```js
import './style.css'
import markdownIt from 'markdown-it'
import copy from './index.md?raw'// added a question mark with the value of raw to import it as text

document.querySelector('#app').innerHTML =
 markdownIt().render(copy)// This code render our data from the markdown to the app
```

Here is the output that shows that the app rendered data from the markdown file.

![Project output](/engineering-education/develop-and-deploy-fast-apps-with-vite-js/output.jpg.jpg)

Vite uses HMR to import other JavaScript modules. Modules are updated individually without reloading other modules to speed up development.

### Conclusion

> Vite is a very fast tool that is easy to set up. Hopefully, you loved working with it as I did. we have learned how to use Vite to create and deploy apps quick. 

Reviewer: ```suggestion
Vite is a very fast tool that is easy to set up. Hopefully, you loved working with it as well as I did. In this tutorial, we have learned how to use Vite to create and deploy apps quick.
